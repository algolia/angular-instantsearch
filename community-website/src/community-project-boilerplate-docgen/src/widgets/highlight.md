---
title: Widgets - Highlight
layout: widget.pug
---

# Highlight

## Description

Renders any attribute from an hit into its highlighted form when relevant.

## Options

| Attributes  | Type     | Description
| -           | -        | -
| `attribute` | `string` | The location of the highlighted attribute in the hit
| `hit`       | `{}`     | The hit object containing the highlighted attribute
| `tagName?`  | `string` | The tag to be used for highlighted parts of the hit

## Code example

```html
<ng-ais-hits>
  <ng-template let-hits="hits">
    <div *ngFor="let hit of hits">
      Hit {{hit.objectID}}:
      <ng-ais-highlight attribute="name" [hit]="hit">
      </ng-ais-highlight>
    </div>
  </ng-template>
</ng-ais-hits>
```
