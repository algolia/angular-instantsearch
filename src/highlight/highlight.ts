import { Component, Input } from "@angular/core";
import { isPlainObject, escape, get } from "lodash-es";
import { bem } from "../utils";

@Component({
  selector: "ais-highlight",
  template: `<span [class]="cx()" [innerHtml]="content"></span>`
})
export class NgAisHighlight {
  @Input() attribute: string;
  @Input() hit: { _highlightResult?: {}; label?: string; highlighted?: string };
  @Input() tagName: string = "em";

  cx = bem("Highlight");

  get content() {
    if (this.attribute === "highlighted") {
      return this.hit.highlighted
        ? this.replaceWithTagName(this.hit.highlighted)
        : this.hit.label;
    }

    if (this.hit.hasOwnProperty("_highlightResult")) {
      const attributeHighlighted = get(
        this.hit._highlightResult,
        this.attribute
      );

      // check that the attributeHighlighted is a string
      if (
        isPlainObject(attributeHighlighted) &&
        typeof attributeHighlighted.value === "string"
      ) {
        return this.replaceWithTagName(attributeHighlighted.value);
      }
    }

    const fallback = get(this.hit, this.attribute);
    if (!fallback) {
      console.warn(
        `Could not find attribute [${
          this.attribute
        }] into hit object, will display an empty string.`
      );

      return "";
    }

    return fallback;
  }

  replaceWithTagName(value: string) {
    return value
      .replace(
        new RegExp("<em>", "g"),
        `<${this.tagName} class="${this.cx("highlighted")}">`
      )
      .replace(new RegExp("</em>", "g"), `</${this.tagName}>`);
  }
}
