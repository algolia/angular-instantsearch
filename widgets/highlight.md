## `<ng-ais-highlight />`

Renders any attribute from an hit into its highlighted form when relevant.

| Attributes      | Type     | Description
| -               | -        | -
| `attributeName` | `string` | The location of the highlighted attribute in the hit
| `hit`           | `{}`     | The hit object containing the highlighted attribute
| `tagName?`      | `string` | The tag to be used for highlighted parts of the hit