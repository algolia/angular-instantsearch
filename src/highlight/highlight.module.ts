import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisHighlight } from './highlight';

@NgModule({
  declarations: [NgAisHighlight],
  entryComponents: [NgAisHighlight],
  exports: [NgAisHighlight],
  imports: [CommonModule],
})
export class NgAisHighlightModule {}
