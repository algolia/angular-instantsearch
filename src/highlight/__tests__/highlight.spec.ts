import { Component } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisHighlightModule } from "../highlight.module";

const render = ({ hit, attribute, tagName = "em" }) => {
  @Component({
    selector: "test-component",
    template: `
      <ais-highlight [attribute]="attribute" [hit]="hit">
      </ais-highlight>
    `
  })
  class TestComponent {
    hit = hit;
    attribute: string = attribute;
    tagName = tagName;
  }

  TestBed.configureTestingModule({
    declarations: [TestComponent],
    imports: [NgAisHighlightModule]
  });

  const fixture = TestBed.createComponent(TestComponent);
  fixture.detectChanges();

  return fixture;
};

describe("highlight", () => {
  it("should highlight strings", () => {
    const fixture = render({
      attribute: "name",
      hit: {
        _highlightResult: {
          name: { value: "<em>foo</em> bar" }
        }
      }
    });
    expect(fixture).toMatchSnapshot();
  });

  it("should highlight nested objects", () => {
    const fixture = render({
      attribute: "parent.name",
      hit: {
        _highlightResult: {
          parent: {
            name: { value: "<em>foo</em> bar" }
          }
        }
      }
    });
    expect(fixture).toMatchSnapshot();
  });

  it("should highlight values in array", () => {
    const fixture = render({
      attribute: "children[0].name",
      hit: {
        _highlightResult: {
          children: [
            {
              name: { value: "<em>foo</em> bar" }
            }
          ]
        }
      }
    });
    expect(fixture).toMatchSnapshot();
  });

  it("should warn when attribute is not found", () => {
    const spy = jest.spyOn(global.console, "warn");
    spy.mockImplementation(() => {});

    const fixture = render({
      attribute: "invalid",
      hit: {
        _highlightResult: {
          name: { value: "<em>foo</em> bar" }
        }
      }
    });
    expect(spy).toHaveBeenCalled();
    expect(fixture).toMatchSnapshot();

    spy.mockReset();
    spy.mockRestore();
  });

  it("should fallback to non highlighted when no match", () => {
    const fixture = render({
      attribute: "name",
      hit: { name: "foo bar" }
    });
    expect(fixture).toMatchSnapshot();
  });

  it("should use `hit.highlighted` if it exists", () => {
    const fixture = render({
      attribute: "highlighted",
      hit: { highlighted: "<em>foo</em> bar" }
    });
    expect(fixture).toMatchSnapshot();
  });
});
