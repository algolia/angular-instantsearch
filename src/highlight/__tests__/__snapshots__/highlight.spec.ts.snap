// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlight should fallback to non highlighted when no match 1`] = `
<test-component
  attribute={[Function String]}
  hit={[Function Object]}
  tagName={[Function String]}
>
  <ais-highlight
    ng-reflect-attribute="name"
    ng-reflect-hit="[object Object]"
  >
    <span
      class="ais-Highlight"
    >
      foo bar
    </span>
  </ais-highlight>
</test-component>
`;

exports[`highlight should highlight nested objects 1`] = `
<test-component
  attribute={[Function String]}
  hit={[Function Object]}
  tagName={[Function String]}
>
  <ais-highlight
    ng-reflect-attribute="parent.name"
    ng-reflect-hit="[object Object]"
  >
    <span
      class="ais-Highlight"
    >
      <em
        class="ais-Highlight-highlighted"
      >
        foo
      </em>
       bar
    </span>
  </ais-highlight>
</test-component>
`;

exports[`highlight should highlight strings 1`] = `
<test-component
  attribute={[Function String]}
  hit={[Function Object]}
  tagName={[Function String]}
>
  <ais-highlight
    ng-reflect-attribute="name"
    ng-reflect-hit="[object Object]"
  >
    <span
      class="ais-Highlight"
    >
      <em
        class="ais-Highlight-highlighted"
      >
        foo
      </em>
       bar
    </span>
  </ais-highlight>
</test-component>
`;

exports[`highlight should highlight values in array 1`] = `
<test-component
  attribute={[Function String]}
  hit={[Function Object]}
  tagName={[Function String]}
>
  <ais-highlight
    ng-reflect-attribute="children[0].name"
    ng-reflect-hit="[object Object]"
  >
    <span
      class="ais-Highlight"
    >
      <em
        class="ais-Highlight-highlighted"
      >
        foo
      </em>
       bar
    </span>
  </ais-highlight>
</test-component>
`;

exports[`highlight should use \`hit.highlighted\` if it exists 1`] = `
<test-component
  attribute={[Function String]}
  hit={[Function Object]}
  tagName={[Function String]}
>
  <ais-highlight
    ng-reflect-attribute="highlighted"
    ng-reflect-hit="[object Object]"
  >
    <span
      class="ais-Highlight"
    >
      <em
        class="ais-Highlight-highlighted"
      >
        foo
      </em>
       bar
    </span>
  </ais-highlight>
</test-component>
`;

exports[`highlight should warn when attribute is not found 1`] = `
<test-component
  attribute={[Function String]}
  hit={[Function Object]}
  tagName={[Function String]}
>
  <ais-highlight
    ng-reflect-attribute="invalid"
    ng-reflect-hit="[object Object]"
  >
    <span
      class="ais-Highlight"
    />
  </ais-highlight>
</test-component>
`;
